packages/assets-controllers/src/AssetsContractController.ts

```diff
@@ -62,137 +67,257 @@ export const SINGLE_CALL_BALANCES_ADDRESS_BY_CHAINID: Record<Hex, string> = {
+  #chainId: Hex;
 
   /**
    * Creates a AssetsContractController instance.
    *
    * @param options - The controller options.
+   * @param options.messenger - The controller messenger.
    * @param options.chainId - The chain ID of the current network.
-   * @param options.onPreferencesStateChange - Allows subscribing to preference controller state changes.
-   * @param options.onNetworkDidChange - Allows subscribing to network controller networkDidChange events.
-   * @param options.getNetworkClientById - Gets the network client with the given id from the NetworkController.
-   * @param config - Initial options used to configure this controller.
-   * @param state - Initial state to set on this controller.
    */
-  constructor(
-    {
-      chainId: initialChainId,
-      onPreferencesStateChange,
-      onNetworkDidChange,
-      getNetworkClientById,
-    }: {
-      chainId: Hex;
-      onPreferencesStateChange: (
-        listener: (preferencesState: PreferencesState) => void,
-      ) => void;
-      onNetworkDidChange: (
-        listener: (networkState: NetworkState) => void,
-      ) => void;
-      getNetworkClientById: NetworkController['getNetworkClientById'];
-    },
-    config?: Partial<AssetsContractConfig>,
-    state?: Partial<BaseState>,
-  ) {
-    super(config, state);
-    this.defaultConfig = {
-      provider: undefined,
-      ipfsGateway: IPFS_DEFAULT_GATEWAY_URL,
-      chainId: initialChainId,
-    };
-    this.initialize();
-    this.getNetworkClientById = getNetworkClientById;
-
-    onPreferencesStateChange(({ ipfsGateway }) => {
-      this.configure({ ipfsGateway });
-    });
-
-    onNetworkDidChange(({ selectedNetworkClientId }) => {
-      const selectedNetworkClient = getNetworkClientById(
-        selectedNetworkClientId,
-      );
-      const { chainId } = selectedNetworkClient.configuration;
+  constructor({
+    messenger,
+    chainId: initialChainId,
+  }: {
+    messenger: AssetsContractControllerMessenger;
+    chainId: Hex;
+  }) {
+    this.messagingSystem = messenger;
+    this.#provider = undefined;
+    this.#ipfsGateway = IPFS_DEFAULT_GATEWAY_URL;
+    this.#chainId = initialChainId;
+
+    this.#registerActionHandlers();
+    this.#registerEventSubscriptions();
+  }
 
-      if (this.config.chainId !== chainId) {
-        this.configure({
-          chainId: selectedNetworkClient.configuration.chainId,
-        });
-      }
-    });
+  // TODO: Expand into base-controller utility function that batch registers action handlers.
+  #registerActionHandlers() {
+    const nonMethodClassProperties = [
+      'constructor',
+      'messagingSystem',
+      'setProvider',
+      'provider',
+      'ipfsGateway',
+      'chainId',
+    ];
+
+    getKnownPropertyNames<keyof this>(Object.getPrototypeOf(this)).forEach(
+      (method) => {
+        if (
+          ((key: keyof this): key is AssetsContractControllerMethodName =>
+            !nonMethodClassProperties.find((e) => e === key) &&
```

Comment:
> Doesn't the typeof this[key] === 'function' check already exclude non-methods? Would it be enough to just exclude setProvider here?

All public methods that are not actions should be explicitly excluded here. Otherwise, the forEach loop will register actions like `AssetsContractController:constructor`. While our types will prevent these from being used, in JavaScript environments, these actions will be accessible at runtime.

This is both an unintended and undesirable extension of the messenger's API, as well as security risk in the form of a ready-made dynamic property access vulnerability (see https://github.com/MetaMask/core/pull/4041 -- especially `constructor` key as attack vector), especially since the action handler `this['constructor'].bind(this)` is bound to the class instance in use.

That said, the variable name `methodsExcludedFromMessenger` is much clearer. Renamed here: f50e5f4c8be72338f83eaf845ff4492b12660196

> Also why use Object.getPrototypeOf(this), why not use this directly?

This results in errors (e.g `"A handler for AssetsContractController:getERC1155BalanceOf has not been registered"`) showing that none of the methods are being registered as action handlers. 

This is because public class methods are defined as sub-properties of the class's `prototype` property, not top-level properties of the class itself, meaning we need to iterate over the prototype object, not `this`.

Reply:
> This is because public class methods are defined as sub-properties of the class's `prototype` property, not top-level properties of the class itself, meaning we need to iterate over the prototype object, not `this`.

Oh, I forgot that `getKnownPropertyNames` doesn't loop over all properties of an object, it only loops over owned properties. Okay, that makes sense now.